TodoMvc example for scalajs-react

elacin opened the pull request:
As mentioned in #1312 - here is what I have come up with.

What do you think?

arthurvr commented on examples/scalajs-react/.gitignore:
```diff
@@ -0,0 +1,3 @@
+.idea
+target
+
```
* [ ] Build up [a decent `.gitignore`](https://github.com/tastejs/todomvc/blob/master/examples/angularjs-perf/.gitignore).
* [ ] The `.idea` should be in your global gitignore.
* [ ] Remove the empty line at the end ;)

arthurvr commented on examples/scalajs-react/index-dev.html:
```diff
@@ -0,0 +1,30 @@
+</script>
+
+</body>
+</html>
```
I'm just running this example and it seems like you forgot the `<h1>todos</h1>` on top of the page.

elacin replied:
Haha good catch, somehow the header style is faint enough that i hadn't seen it was even missing! :)

arthurvr commented on examples/scalajs-react/index-dev.html:
```diff
@@ -0,0 +1,30 @@
+<footer class="info">
+    <p>Double-click to edit a todo</p>
+
+    <p>Created by <a href="http://github.com/elacin/">elacin</a></p>
```
Use your full name instead of your user name.

arthurvr commented on examples/scalajs-react/index-dev.html:
```diff
@@ -0,0 +1,30 @@
+<html lang="en" data-framework="scalajs-react">
+<head>
+    <meta charset="utf-8">
+    <title>scalajs-react (dev) • TodoMVC</title>
```
Remove the `(dev)` thing

arthurvr commented on examples/scalajs-react/index-dev.html:
```diff
@@ -0,0 +1,30 @@
+</footer>
+
+<script src="node_modules/todomvc-common/base.js"></script>
+<script type="text/javascript" src="generated/todomvc-jsdeps.js"></script>
```
`type="text/javascript"` is moot.

arthurvr commented on examples/scalajs-react/index-dev.html:
```diff
@@ -0,0 +1,30 @@
+
+<script src="node_modules/todomvc-common/base.js"></script>
+<script type="text/javascript" src="generated/todomvc-jsdeps.js"></script>
+<script type="text/javascript" src="target/scala-2.11/todomvc-fastopt.js"></script>
```
`type="text/javascript"` is moot.

arthurvr commented on examples/scalajs-react/index-dev.html:
```diff
@@ -0,0 +1,30 @@
+<script type="text/javascript" src="generated/todomvc-jsdeps.js"></script>
+<script type="text/javascript" src="target/scala-2.11/todomvc-fastopt.js"></script>
+
+<script type="text/javascript">
```
`type="text/javascript"` is moot.

elacin replied:
The same could be said for the entire index page, It's there just for bootstrapping purposes - the point of this whole ordeal was exactly to avoid writing this stuff myself, hehe.

Anyways, I'll be removing that type, then

arthurvr replied:
I don't understand how this is related to bootstrappping purposes. If this is bootstrapped by some kind of tool that tools needs to be fixed there as well.

`type="text/javascript"` is the default on every script tag so it shouldn't be there. You don't write any other default attribute as well. That's how we keep codebases/code style clean ;)

arthurvr commented on examples/scalajs-react/index-dev.html:
```diff
@@ -0,0 +1,30 @@
+
+<script type="text/javascript">
+  ((typeof global === "object" && global &&
+         global["Object"] === Object) ? global : this)["todomvc"]["Main"]().main();
```
Please follow [our JavaScript code style](https://github.com/tastejs/todomvc/blob/master/codestyle.md) in this `<script>`. Also many of these files have the wrong indentation, we use tabs, no spaces :)

elacin replied:
About this line - It's a random bootstrapping incantation that I copied from the docs of the library I'm using, so I'm sure there is a reason why its written like that. Unless you have a good propsition for how to simpify it and keep it working?

arthurvr replied:
I don't understand how you would need to "simplify" this. I didn't complain about the actual code but about the code style, the code shouldn't change: single-quotes, tab indentation, ...

elacin replied:
aha! I didn't understand what you meant was wrong with it originally, in my eyes it's just an ugly line of code which does magic things to start the code i actually wrote, hehe. 

I'll restyle it

arthurvr commented on examples/scalajs-react/index-dev.html:
```diff
@@ -0,0 +1,30 @@
+
+</script>
+
+</body>
```
Please indent inside of the `<html>` tag.

arthurvr commented on examples/scalajs-react/index-dev.html:
```diff
@@ -0,0 +1,30 @@
+    <link rel="stylesheet" href="node_modules/todomvc-app-css/index.css">
+</head>
+<body>
+<section id="todoapp" class="todoapp"></section>
```
Please indent inside of the `<body>`.

arthurvr commented on examples/scalajs-react/index-dev.html:
```diff
@@ -0,0 +1,30 @@
+         global["Object"] === Object) ? global : this)["todomvc"]["Main"]().main();
+
+</script>
+
```
remove this blank line.

arthurvr commented on examples/scalajs-react/index-dev.html:
```diff
@@ -0,0 +1,30 @@
+<script type="text/javascript">
+  ((typeof global === "object" && global &&
+         global["Object"] === Object) ? global : this)["todomvc"]["Main"]().main();
+
```
remove this blank line.

arthurvr commented on examples/scalajs-react/index-dev.html:
```diff
@@ -0,0 +1,30 @@
+<section id="todoapp" class="todoapp"></section>
+<footer class="info">
+    <p>Double-click to edit a todo</p>
+
```
remove this blank line.

arthurvr commented on examples/scalajs-react/index-dev.html:
```diff
@@ -0,0 +1,30 @@
+    <p>Double-click to edit a todo</p>
+
+    <p>Created by <a href="http://github.com/elacin/">elacin</a></p>
+
```
remove this blank line.

arthurvr commented:
Hey-hey @elacin! Thanks for this PR! You rock! Sorry for all the nitpicking. Can you please run all our tests? Many of them seem to fail for me.

arthurvr commented on examples/scalajs-react/readme.md:
```diff
@@ -0,0 +1,69 @@
+ which references the files generated from `fastOptJS`.
+
+ 
+Happy hacking! :)
```
About this whole file, try to make the format consistent with the other examples readme files.

arthurvr commented on learn.json:
You linked SO twice.

arthurvr commented on examples/scalajs-react/src/main/scala/todomvc/types.scala:
```diff
@@ -0,0 +1,32 @@
+
+  def values = Seq[TodoFilter](All, Active, Completed)
+}
+
```
Let's remove that blank line.

elacin commented:
Thanks for the comments, @arthurvr , I have a bit time to update it now.

Travis fails because jscsrc spends forever (and eventually complains when) linting my generated js, which is quite understandable. I see there is an entry for exclusions in .jscsrc, can i use that?
```diff
-       "excludeFiles": ["**/node_modules/**", "**/bower_components/**"],
+       "excludeFiles": ["**/node_modules/**", "**/bower_components/**", "**/generated/**"],
```

arthurvr commented:
@elacin Good point. We only just started running JSCS in CI. Adding it to the ignores there sounds like the right thing to do.

elacin commented:
The tests «works on my machine» for now, do you have any idea what's wrong, @arthurvr ?

I'm still missing bringing readme.md into line with the others, but I'm going to have to leave that for tomorrow i think. will rebase everything into come commit once were good to go with the tests :)

arthurvr commented:
@elacin Sorry, my bad. Let's use `todomvc-app-css` 2.0. The tests don't work with that yet but I normally that should be fixed soonish.

elacin commented:
ok, im reverting that commit, then

arthurvr commented on examples/scalajs-react/package.json:
```diff
@@ -0,0 +1,7 @@
+	"todomvc-app-css": "^2.0.1",
+	"todomvc-common": "^1.0.2"
+  }
+}
```
The indentation in this file is screwed up.

elacin replied:
fixing, autoconversion failed for that file

arthurvr commented:
@elacin @tastejs/todomvc do we know anybody to review the actual scala code? I'm definitely not the right one to do that.

arthurvr commented on examples/scalajs-react/sbt:
```diff
@@ -0,0 +1,543 @@
+shouldFilter () { [[ -f ~/.sbtignore ]] && ! egrep -q '\b(shell|console|consoleProject)\b' <<<"${residual_args[@]}"; }
+
+# run sbt
+if shouldFilter; then mainFiltered; else main; fi
```
I don't want random third-party files living in the repo which aren't managed by any package manager. Is this one?

elacin replied:
Yes, i guess it is. It is an sbt launcher for people without sbt installed so they can easily build the example by following the readme. Sure i can remove that

arthurvr commented:
Something seems to be going wrong with the styling of the control buttons at the bottom of the app.

elacin commented:
i'll try to get some knowledgeable people to review my scala code.

elacin commented:
i agree something seems to be a bit off with the filters, im at least missing some padding on the right side of the clear all-button. i dont seem to be missing any classes i think, can it be a todomvc-app-css 2 problem? deleted the launcher. still readme file pending

elacin commented:
So, im working on integrating changes from reviewers on the scala code, think its coming along quite fine now.

Can we make an exception for the tabs instead of spaces-rule for scala files? I guess it's a javascript thing anyway, and I don't really see that is has to apply for non-js languages which use spaces.

arthurvr commented:
> Can we make an exception for the tabs instead of spaces-rule for scala files? I guess it's a javascript thing anyway, and I don't really see that is has to apply for non-js languages which use spaces.

Yeah, for scala that's alright I guess. Though make sure to edit the `.editorconfig`.

elacin commented:
Right, it's ready for a another look, @arthurvr. 

Since last time:
* script tags in html cleaned up
* removed some empty lines
* added .editorconfig for the spaced scala files
* rewrote readme in the style of the others (copied a few links from the other react examples, i guess thats fine - also added links)
* removed one stackoverflow link from learn.json
* fixed layout issue with the footer

As for the scala code, i got some feedback from the creator of the library I'm using ( see https://github.com/japgolly/scalajs-react/issues/54#issuecomment-109664142 ), and fixed most of the things he commented except two:
- there are two stateful components. It can be argued fewer places is better, but I'm keeping it as is because i like the separation of concerns it buys me. This is also very much in line with how it was done for the plain react example here in todomvc.
- a suggestion to simplify some code with lenses, which i didn't do because this code already has enough new concepts for newcomers.

Tests are still failing, but keep in mind that once i ported back to the old css they «worked on my machine» :)

arthurvr commented on examples/scalajs-react/index.html:
```diff
@@ -0,0 +1,23 @@
+		<!--run application-->
+		<script src="generated/todomvc-launcher.js"></script>
+	</body>
+</html>
```
needs a final newline

passy replied:
^ Still needs this one.

arthurvr commented on examples/scalajs-react/readme.md:
```diff
@@ -0,0 +1,90 @@
+> _[Scalajs-React - https://github.com/japgolly/scalajs-react/](https://github.com/japgolly/scalajs-react/)_
+
+
+##Learning Scala
```
Please have a space after the `##`

arthurvr commented on examples/scalajs-react/readme.md:
```diff
@@ -0,0 +1,90 @@
+* [React on GitHub](https://github.com/facebook/react)
+* [Support](http://facebook.github.io/react/support.html)
+
+More links under the other React TodoMVC examples [react example](../react) 
```
React with a capital letter

arthurvr commented on examples/scalajs-react/package.json:
```diff
@@ -0,0 +1,7 @@
+		"todomvc-app-css": "^2.0.1",
+		"todomvc-common": "^1.0.2"
+	}
+}
```
Use 2 spaces indentation in `package.json`. Our editorconfig states so.

arthurvr commented on examples/scalajs-react/readme.md:
````diff
@@ -0,0 +1,90 @@
+You can for example start one like this
+
+```
+python3 -m http.server 4000
````
What's wrong with `python -m SimpleHTTPServer 8000` which works by default?

samccone replied:
or just `python -m SimpleHTTPServer` :)

elacin replied:
i went with the one with explicit port number, i always like to see where i can change that myself.

arthurvr commented on examples/scalajs-react/project/Build.scala:
```diff
@@ -0,0 +1,43 @@
+      crossTarget in (Compile, packageMinifiedJSDependencies) := generated,
+      artifactPath in (Compile, fastOptJS) := ((crossTarget in (Compile, fastOptJS)).value / ((moduleName in fastOptJS).value + "-opt.js"))
+    )
+}
```
newline

elacin commented:
For next time make sure to mention the existence of for example `editorconfig-tools` - could have saved us both some time if i knew! :)

Current status, after changing to `root=false` in this `.editorconf`:
```
$ editorconfig-tools check `git ls-tree -r scalajs-react-rebased|cut -f2|grep -v generated`|grep -v max_line_length
readme.md failed indent_char: found setting '  ', should be '	'
readme.md failed trim_trailing_whitespace on line 0: found setting 'true', should be 'false'
```
it complains for every file that `max_line_length` is not set, I didn't want to impose anything on the whole project to shut it up. 

I don't even really understand those two, but I guess we're good on the whitespace issues?

arthurvr commented:
> For next time make sure to mention the existence of for example editorconfig-tools - could have saved us both some time if i knew! :)

If you just had used the `editorconfig` plugins in your editor that wouldn't have been any problem.

> it complains for every file that max_line_length is not set, I didn't want to impose anything on the whole project to shut it up.

Just don't use that tool, and instead use your editor (where it's meant for). Our editorconfig is perfectly valid.

elacin commented:
yeah, sure. 

Rebased it again now, do you have any outstanding issues here?

passy commented:
This is looking great! Really impressive DSL you've embedded into Scala for this.

One thing, that is probably not related to this specific example, but a bit worrying: Compared to the JS React implementation, this one doesn't actually diff the DOM correctly and makes way too many DOM updates if a single element changes. Here for example is the behavior if you tick an item:

![screenshot 2015-06-11 12 22 53](https://cloud.githubusercontent.com/assets/9906/8105892/ecd5ece8-1034-11e5-9bb6-4c8cebfb0352.png)

The JS example does the same in four operations, whereas the Scala.js+React one needs 10 and would grow with the size of the list. Any idea why that is?

arthurvr commented on examples/scalajs-react/.editorconfig:
```diff
@@ -0,0 +1,6 @@
+# editorconfig.org
```
Let's remove this comment.

arthurvr commented on examples/scalajs-react/generated/todomvc-launcher.js:
```diff
@@ -0,0 +1,2 @@
+((typeof global === "object" && global &&
+         global["Object"] === Object) ? global : this)["todomvc"]["Main"]().main();
```
Still we want tab indentation in JS files. Also, as I told you earlier please use our javascript style guide.

elacin replied:
Yeah i did first rewrite it like we agreed. 

The reason why it looks like this again now, is that it is now generated (look at the path) by the sbt scala.js plugin - in Build.sbt:
```diff
+      /* create javascript launcher. Searches for an object extends JSApp */
+      persistLauncher                := true,
```

I guess we can agree that generated code is not subject to the style guide? :)

arthurvr replied:
Yeah, sure.

elacin commented:
Hah, great observation there, @passy . There are some tools to debug performance, I'll have a look at it and see if i can figure out why we get those extra DOM updates! :)

elacin commented:
So, i worked a bit more on this, then. vdom diffing now works like it's supposed to. I was a bit naive and `createClass`ed every component for every redraw without realizing it, so react rightly refused to update them. 

Also wrote a few comments that i figured would be useful.

Again, think we should be about good to go :)

arthurvr commented on examples/scalajs-react/index.html:
```diff
@@ -0,0 +1,23 @@
+		</footer>
+		<script src="node_modules/todomvc-common/base.js"></script>
+		<script src="generated/todomvc-jsdeps.js"></script>
+		<!--include application-->
```
`<!-- include application -->` notice the spacing

arthurvr commented on examples/scalajs-react/index.html:
```diff
@@ -0,0 +1,23 @@
+		<script src="generated/todomvc-jsdeps.js"></script>
+		<!--include application-->
+		<script src="generated/todomvc-opt.js"></script>
+		<!--run application-->
```
`<!-- run application -->`, notice the spacing

arthurvr commented on examples/scalajs-react/readme.md:
```diff
@@ -0,0 +1,90 @@
+and [React-Backbone example](../react-backbone)
+
+
+## Learning Scala.js
```
Include a blank line after this for consistency

arthurvr commented on examples/scalajs-react/readme.md:
```diff
@@ -0,0 +1,90 @@
+* [Tutorial](http://www.scala-js.org/doc/tutorial.html)
+
+
+## Learning Scalajs-React
```
Include a blank line after this for consistency

arthurvr commented on examples/scalajs-react/readme.md:
```diff
@@ -0,0 +1,90 @@
+* [Documentation/Extras](https://github.com/japgolly/scalajs-react/blob/master/extra/README.md)
+
+
+## Running
```
I wonder if this block is really needed. It's common knowledge and not really specific to this example

arthurvr commented on examples/scalajs-react/readme.md:
```diff
@@ -0,0 +1,90 @@
+
+After installation, you start `sbt` like this:
+
+`sbt`
```
this should be a code block (like line 60). Also include a dollar sign (`$`) to clarify the fact that this is a command.

arthurvr commented on examples/scalajs-react/readme.md:
```diff
@@ -0,0 +1,90 @@
+
+After `sbt` has downloaded the internet, you can compile once like this:
+
+`fastOptJS` 
```
this should be a code block (like line 60). Also include a dollar sign ($) to clarify the fact that this is a command.

arthurvr commented on examples/scalajs-react/index.html:
```diff
@@ -0,0 +1,23 @@
+		<link rel="stylesheet" href="node_modules/todomvc-app-css/index.css">
+	</head>
+	<body>
+		<section id="todoapp" class="todoapp"></section>
```
Why is the `id` still here? The class should be enough.

elacin replied:
yeah, good point - it was left over and i used it to find where in the dom i would render, changed it to lookup by class instead

arthurvr commented:
@samccone Could you test this using leaky? 

-

We recently landed updates to the test suite and everything passes! :tada: Thanks for your awesome work!

elacin commented:
yeah thanks, great that you were able to update the tests so theyre working, too. Think this is turning out quite well. 

I'll push a final, rebased commit soon with some updated comments that i feel would be useful now that i looked at it again.

elacin commented:
there, ready again on my side. Now uses minimized js deps, otherwise pretty much the same as before

elacin commented:
tests failed there for a second, then succeeded with the same code, so something might not be entirely stable yet

arthurvr commented:
> Now uses minimized js deps,

Why? We don't want that: https://github.com/tastejs/todomvc/issues/378.

samccone commented:
@elacin you are correct there does seem to be some more flake
https://github.com/tastejs/todomvc/issues/1378

elacin commented:
oh, i thought that would be preferable and considered it an oversight to use unminified. reverting that then

elacin commented:
@arthurvr can we merge this?

arthurvr commented:
> @arthurvr can we merge this?

Hey there, I'll try to give this another full review ASAP. I'm just back from a vacation.

arthurvr commented:
@samccone Can you check this using leaky?

samccone commented:
@elacin super sorry for the slow turn on this one, would you mind rebasing this one :)

passy commented on examples/scalajs-react/src/main/scala/todomvc/types.scala:
```diff
@@ -0,0 +1,29 @@
+  def editable = UnfinishedTitle(value)
+}
+
+case class UnfinishedTitle(value: String){
```
Nit, but could we get a space after the closing bracket? There are a few other places in the diff, too.

passy commented on examples/scalajs-react/src/main/scala/todomvc/CTodoItem.scala:
```diff
@@ -0,0 +1,100 @@
+      $.state.editText.validated.fold($.props.onDelete)($.props.onUpdateTitle)
+
+    /**
+     * Its OK to make these into `val`s as long as they dont touch state.
```
*It's

passy commented on examples/scalajs-react/src/main/scala/todomvc/CTodoItem.scala:
```diff
@@ -0,0 +1,100 @@
+        case _              => None
+      }
+
+    val editFieldChanged: ReactEventI => IO[Unit] =
```
Loving the explicit effects. <3

elacin replied:
Yeah, this is definitely one of the things that I'm very excited about with this technology too :)

passy commented on examples/scalajs-react/src/main/scala/todomvc/CTodoList.scala:
```diff
@@ -0,0 +1,148 @@
+  /**
+   * These specify when it makes sense to skip updating this component (see comment on `Listenable` below)
+   */
+  implicit val r1 = Reusability.fn[Props]((p1, p2) ⇒ p1.currentFilter == p2.currentFilter)
```
The unicode arrows are fun, but inconsistent with the rest of the app. Could we stick with one please?

elacin replied:
yeah, i meant to take them all out, but forgot to tell my IDE. gone now

passy commented on examples/scalajs-react/src/main/scala/todomvc/CTodoList.scala:
```diff
@@ -0,0 +1,148 @@
+     *           If not, rerendering will prohibit the second from having its effect.
+     *           For this example, the current `State` contains both `editing` and the list of todos.
+     */
+    def editingDone(cb: OpCallbackIO = js.undefined): IO[Unit] =
```
Just curious, why does this take an "untyped" JS value rather than an Option?

elacin replied:
It's actually perfectly typed - `OpCallbackIO` is a type alias defined like this:
```scala
type OpCallbackIO = js.UndefOr[scalaz.effect.IO[Unit]]
```
In other words, it is an optional callback. `js.undefined` can fill that hole in the same way as does `None` for the `Option` type - it's the instance of `js.UndefOr[Nothing]`.

To a pretty big extent we use `js.UndefOr[A]` instead of `Option[A]` in scala.js. It has `Option`-like semantics, but instead uses `js.undefined` as a `null` value instead of runtime wrapper objects. This is done primarily for integration with javascript components. A nice added bonus in the context of react and scala.js is that it also allows us to compare optionals by reference - which you frequently want to do in `shouldComponentUpdate`-

passy replied:
Great explanation. Thank you, makes perfect sense! :+1:

passy commented on examples/scalajs-react/src/main/scala/todomvc/TodoModel.scala:
```diff
@@ -0,0 +1,50 @@
+    var todos = Seq.empty[Todo]
+
+    def mod(f: Seq[Todo] => Seq[Todo]): IO[Unit] =
+      IO{
```
Space before `{`

passy commented on learn.json:
https, please

passy commented:
The code itself looks great. Just a few very minor remarks. Thanks for this!

elacin commented on learn.json:
fixed

elacin commented:
yeah dont worry @samccone - it's not like it's terribly urgent to get it up. Thanks for the comments @passy , quite sure i took care of all of them now. Also rebased on top of master

elacin commented:
So we failed tests on a check that the clean all-button is not visible, which i fixed now. You might want to consider the case that the button doesnt exist in the DOM as a success case. I felt it was unnecessary to create a button in `render()` with visibility.hidden just to pass a test.

arthurvr commented:
Mind fixing the merge conflict? Thanks :)

elacin commented:
There you go, @arthurvr

arthurvr commented:
Thanks. So, then I don't think there's anything outstanding anymore. Tests pass too. LGTM. @passy @samccone Would appreciate some extra eyes on this though.

elacin commented:
Great @arthurvr , looking forward to getting it merged :) I pushed a new version again now, seems i had introduced a react warning to the console, so i figured it was worth fixing.

samccone commented:
Yep yep will take a look and run it through the paces

samccone commented:
LGTM passes the memory test with flying colors :)

arthurvr commented:
Thanks Sam. Going to let @passy take a final look and then land this one. Thanks for the hard work @elacin!

passy commented:
Aaaand landed! Thanks for the amazing work, @elacin and @arthurvr and @samccone for helping! :sparkles: 

![](https://media.giphy.com/media/3oEduPiS1VU9azQb9m/giphy.gif)
